pkg/validation/options.go

```diff
@@ -427,6 +433,41 @@ func parseJwtIssuers(issuers []string, msgs []string) ([]jwtIssuer, []string) {
 	return parsedIssuers, msgs
 }
 
+func parseWhitelistIPs(ipWhitelistStrs []string, msgs *[]string) []*net.IPNet {
```

Comment:
I wonder if, instead of parsing and returning the list here, it's better to just check they're valid and return messages. Then pass the unparsed strings into `NewNetSet` and have that do all the parsing and return an error there if there's a problem (which they're shouldn't be assuming validation passes). 

If you do it this way around, then you don't need to add the private field to the options struct and get/set that. Plus we are just doing validation in the validation rather than validation and setting up for later.

It would also help writing tests for the `NewNetSet` in my opinion, as the input you give would just be strings (identical to what the users may put in) so it would be easier to read/reason about for developers or users browsing the code

What do you think?

Reply:
I have two problems with this idea, although it was initially my approach.

1. In order to properly validate, it's reasonably required to parse the IPs.  This means doing one round of parsing in the options (but ignoring the result) and then parsing again at the whitelist creation.  More over, it has to be done in a somewhat specific way, in which we infer a `/32` or `/128` netmask for a bare IP.  This results in somewhat complicated logic that we must ensure is duplicated perfectly in two places to avoid any hard to track down problems that might show up in edge cases.
1. It requires putting option-parsing and oauth2-proxy global needs into NetSet, making it less of an abstract use-for-whatever-you-like data structure and a more specific ad hoc data structure that would be more difficult to make fill other needs.

Both of these problems can be sidestepped by creating a separate routine that parses IPs for this need.  But that carries with it it's own problems, in particular decreased brevity and an otherwise unnecessary and useless abstraction.

It certainly can be done, but in each case it comes with its own set of downsides.

Reply:
> In order to properly validate, it's reasonably required to parse the IPs. This means doing one round of parsing in the options (but ignoring the result) and then parsing again at the whitelist creation. More over, it has to be done in a somewhat specific way, in which we infer a /32 or /128 netmask for a bare IP. This results in somewhat complicated logic that we must ensure is duplicated perfectly in two places to avoid any hard to track down problems that might show up in edge cases.

I'm not concerned about performance of parsing the lists twice, they're only done at startup so once the project has started there should be no impact on response times, and I don't think it's likely to take that long to parse anyway, WDYT? 
As for duplicating the logic, I had imagined there would be some publicly accessible parsing function in the IP package that could be called here `_, err := ip.Parse(...)`, which would be called exactly the same way in `NewNetSet`

> It requires putting option-parsing and oauth2-proxy global needs into NetSet, making it less of an abstract use-for-whatever-you-like data structure and a more specific ad hoc data structure that would be more difficult to make fill other needs.

I had imaged we would just pass a list of IPs and letting it do the parsing, would that really make it less useful? I'm not sure where else we would use it tbh, but we could also have two constructors, one for constructing from net.IP as we have now, and one that Parses strings and returns the NetSet, WDYT?

> Both of these problems can be sidestepped by creating a separate routine that parses IPs for this need. But that carries with it it's own problems, in particular decreased brevity and an otherwise unnecessary and useless abstraction.

I don't follow why this has problems, would extracting the parsing into it's own function not just be copying what we already have here but making it public?

Reply:
> As for duplicating the logic, I had imagined there would be some publicly accessible parsing function in the IP package that could be called here `_, err := ip.Parse(...)`, which would be called exactly the same way in `NewNetSet`

It's this bit here. We rely on slightly more complicated logic: we add `/32` and `/128` netmasks if no netmask is provided.  It's this behaviour that I don't want to have to duplicate.

> I don't follow why this has problems, would extracting the parsing into it's own function not just be copying what we already have here but making it public?

If you're happy with this strategy, I'll implement it as our solution to this problem.

Reply:
I think I'm happy with it.

Basically, long term I would like Validate to be purely validation and not modify any of the settings. So all it will do is do preflight checks for the set up of the rest of the proxy, if that means calling functions that parse things and checking they don't return an error, then that is ok in this model I think.

You could argue what's the point of validating at all, why not just let the set up fail later, which is one option, but I'd prefer to try lots of setups and have them all fail at once so we can return the most helpful messages to users, rather than sending one after another after another as users resolve issues.

Does that make sense?

Reply:
> long term I would like Validate to be purely validation and not modify any of the settings

Yes I totally understand, and agree it's admirable, and quite practical in this particular case.  But I'd warn that sometimes validating might be inherently destructive or expensive.  I'm not sure if we'll ever see this extreme in oauth2-proxy, but it certainly could be the case.

> I think I'm happy with it.

I'll go for it, it seems like the best compromise between correctness of validation (100% correctness), complexity (one unnecessary abstraction).

Reply:
> Yes I totally understand, and agree it's admirable, and quite practical in this particular case. But I'd warn that sometimes validating might be inherently destructive or expensive. I'm not sure if we'll ever see this extreme in oauth2-proxy, but it certainly could be the case.

Yeah I appreciate this, and at the end of the day, it's a bit of an experiment, it could well be a mistake and in that case, we will have to undo it, hopefully we will find out soon enough 😅 

> I'll go for it, it seems like the best compromise between correctness of validation (100% correctness), complexity (one unnecessary abstraction).

Thank you, I appreciate you working with me on this to improve it, I realise it's been going on for a while, I think we are very close now